components/base-rest-api/packages/services/lib/token-swapper-service.js

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import _ from 'lodash';
+import { Service } from '@aws-ee/base-services-container';
+import jwtDecode from 'jwt-decode';
+import { newInvoker } from './authentication-providers/helpers/invoker';
+
+const settingKeys = {
+  dbValidTokens: 'dbValidTokens',
+};
+class TokenSwapperService extends Service {
+  constructor() {
+    super();
+    this.boom.extend(['invalidToken', 403]);
+    this.dependency(['dbService', 'authenticationProviderConfigService']);
+  }
+
+  async init() {
+    this.invoke = newInvoker(this.container.find.bind(this.container));
+  }
+
+  async swap({ token, uid }) {
+    const dbService = await this.service('dbService');
+    const dbValidTokens = this.settings.get(settingKeys.dbValidTokens);
+
+    if (!uid) {
+      throw this.boom.badRequest('Empty uid', true);
+    }
+    let payload;
+    try {
+      payload = jwtDecode(token);
+    } catch (error) {
+      throw this.boom.invalidToken('Invalid Token', true).cause(error);
+    }
+    // New token should not be expired
+    const ttl = _.get(payload, 'exp', 0);
+    if (ttl === 0) {
+      throw this.boom.invalidToken('Expired Token', true);
+    }
+    // If token is the same as the token in the dbValidToken then
+    // swap is not required.
+    const { isSameToken } = await this.revokeValidToken({ token, uid });
+    if (isSameToken) {
+      return;
+    }
+    const record = { uid, token, ttl };
```

Comment:
We should not store a raw access_token in DDB that token could be re-used by system admin to impersonate a user. Consider hashing the value and storing that a good article explaining some options: https://www.vaadata.com/blog/how-to-securely-store-passwords-in-database/